### DISCUSSION TITLE
Add support to SML Modbus multiple meters on one serial interface

### DISCUSSION THREAD

**Discussion by kcbam** [arendst/Tasmota#15320] [Ideas]
### PROBLEM DESCRIPTION
Tasmota Modbus SML Module does not support multiple meters on one serial interface. For every meter that is configured in the script, tasmota creates a new object of the serial interface. It would be reasonable to configure one serial interface (rs485 module) first, then configuring the meters on this interface. At the moment it is not possible to connect several Modbus devices in parallel to one serial interface.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): _____
- [x] Tasmota binary firmware version number used: _____
  - [x] Pre-compiled
  - [x] Self-compiled
- [x] Flashing tools used: _____
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by gemu2015** [comment]
Check the docs
There are examples of several meters on one bus!
You define only one meter (meter = bus)

**Reply by kcbam** [reply]
This is not a good solution, because in this case you have only one MQTT topic. I have several (identical) energymeters with about 20 parameters each.

**Comment by gemu2015** [comment]
if you insist in having different topics per MODBUS try the following solution:
define your meter with MODBUS topic \*
this disables default MQTT
then create your own MQTT for every meter like this

\>J
,"mb1":{
"total_in":%2sml[1]%,
"total_out:%2sml[2]%}
,"mb2":{
"total_in":%2sml[3]%,
"total_out:%2sml[4]%}

sml[x] x being the decode line of your decoder section, the '2' being the number of decimals

you must also enable
#define USE_SML_SCRIPT_CMD
#define USE_SCRIPT_JSON_EXPORT

your setup will produce heavy load. consider using an ESP32
you may need a file system to use larger script sizes because your decoder will get large, see scripting docs
you will have to increase number of SML variables, see SML docs.

you may save script space by replacing decoder text with this option, see docs and examples
#define SML_REPLACE_VARS

**Reply by kcbam** [reply]
I've tested this and unfortunately found that it doesn't work either. The problem is that you create a single MQTT topic with a very long JSON string. The message is then incomplete, probably because the memory is full. It would be better if each meter got its own MQTT topic and was processed one after the other.

**Reply by gemu2015** [reply]
to queue your MQTT messages you can do the following:

in >D define a counter  jcnt=0

in >J call a subroutine %=#mqout

you now can generate MQTT under program control
rapp means script cmd respond_append

#mqout
if jcnt%2==0 {
rapp ,"mb1":{"total_in":%2sml[1]%,"total_out:%2sml[2]%}
} else {
rapp ,"mb2":{"total_in":%2sml[3]%,"total_out:%2sml[4]%}
}
jcnt+=1

or you may simply send your MQTT by ->publish

**Comment by kcbam** [comment]
Thanks for the suggestion, I will look into it soon. 
But wouldn't it make sense to adapt the code? In my opinion, it is currently a bug. If I create several Modbus meters, only the parameters of one are read. In the code, when creating the serial port object, it should be checked if there is already an object with the pins.

**Reply by gemu2015** [reply]
no this won't work (for the receiver)
since the serial buffers are polled meter by meter the first meters serial buffer would consume all the data and the decoder of meters 2 and up wouldn't decode anything
this would require a different specialized MODBUS adapted architecture of the driver

**Comment by jkwim** [comment]
I have a similar requirement to hook up more than one meter in parallel.

So I configured like:

```
>M 2  
+1,3,M,0,9600,DDM1,1,5,01040000
1,010404ffffffff@i0:1,Voltage,V,DDM1_Voltage,2
.
.
.
+2,3,M,0,9600,DDM2,1,5,02040000
2,020404ffffffff@i0:1,Voltage,V,DDM2_Voltage,2 
.
.
.
#
```

In console dump more I can see the messages being sent for both devices and responses received.

Output picks only one device's response.

Is this how it is supposed to work? Deconding only one devices responses?

**Reply by jkwim** [reply]
On a second thought I wonder whether I can define two devices in to a single meter definition and combine the addresses as follows (since Tasmota does not know which device responds):

```
>M 1  
+1,3,M,0,9600,DDM,1,5,01040000,02040000
1,010404ffffffff@i0:1,Voltage,V,DDM1_Voltage,2
1,020404ffffffff@i0:1,Voltage,V,DDM2_Voltage,2 
#
```

Note the two different device IDs 01 & 02

**Reply by jkwim** [reply]
> On a second thought I wonder whether I can define two devices in to a single meter definition and combine the addresses as follows (since Tasmota does not know which device responds):
> 
> ```
> >M 1  
> +1,3,M,0,9600,DDM,1,5,01040000,02040000
> 1,010404ffffffff@i0:1,Voltage,V,DDM1_Voltage,2
> 1,020404ffffffff@i0:1,Voltage,V,DDM2_Voltage,2 
> #
> ```
> 
> Note the two different device IDs 01 & 02

This setup works fine for me.

**Comment by tmycgn** [comment]
I have a similar question. I am trying to operate the following 2 meters at the same time. They work individually, but not together. Is this not possible or am I simply doing something wrong? I would be very happy for any help! 😊

```
>D
>B
->sensor53 r
>M 2
+1,3,kN2,0,1200,Kamp,1,5,3F1004003C005600570059,3F1003004A0044007B
+2,4,m,0,19200,Trovis,5,2,r0A030009000E,r0A03001C0004,0A03006A
1,3F10003Ckstr@i0:1000,Waerme,MWh,Waerme,3
2,0A031CSSss@i0:10,Aussen,C,Outside,1
#

```

**Reply by gemu2015** [reply]
@tmycgn 
you will not have any luck with an esp8266 here because it only has 1 hardware serial. with any ESP32 there should be no problem at all.

**Reply by tmycgn** [reply]
@gemu2015 Thank you for your answer, it is an ES32 C3 (Mini). Would this one do the job?
If yes, do you have any idea what the problem could be? Tasmota 13.4.0 (tasmota32) is installed.

![image](https://github.com/user-attachments/assets/30495429-2d05-472c-a68d-355b1ae2e039)

**Reply by gemu2015** [reply]
this one has 2 hardware UARTS but due to a bug in core2 you could only use one of it.
this bug has been removed with core3. simply update to the most recent firmware.

**Reply by tmycgn** [reply]
Thanks for the info, you're helping me really a lot. Can you tell me from which Tasmota version the bug was fixed?

**Reply by tmycgn** [reply]
**UPDATE:**
Small smart home world...

I have found you @gemu2015 and Hichi here, both of whom I am in contact with. 😄

[https://forum.creationx.de/forum/index.php?thread/4242-2-lesek%C3%B6pfe-wollen-nicht-gleichzeitig/&pageNo=2](https://forum.creationx.de/forum/index.php?thread/4242-2-lesek%C3%B6pfe-wollen-nicht-gleichzeitig/&pageNo=2)

It seems we are all 3 trying to enable the same hardware i.e. Hichi's Wifi2 Smart Meter. 

If the solution is now Tasmoto 14 with Core 3, it would be great if there was a new customized firmware for the device. If someone has already created this and could make it available at this point, I would of course be very happy. ✌️

**UPDATE2:**

Found the right firmware with Hichi's help here, works great. Thanks to all!

[https://ottelo.jimdofree.com/stromz%C3%A4hler-auslesen-tasmota/#Downloads](https://ottelo.jimdofree.com/stromz%C3%A4hler-auslesen-tasmota/#Downloads)
